src/Build.UnitTests/Evaluation/Evaluator_Tests.cs

```diff
@@ -4385,6 +4385,38 @@ public void ThrownInvalidProjectExceptionProperlyHandled()
             }
         }
 
+        /// <summary>
+        /// Tests that an import, target, or task with a condition that contains an error but it short circuited does not fail the build.  This can happen when you have a condition like:
+        /// 'true' == 'false' AND '$([MSBuild]::GetDirectoryNameOfFileAbove($(NonExistentProperty), init.props))' != ''
+        /// 
+        /// The first condition is false so the second condition is not evaluated.  But in some cases we double evaluate the condition to log it.  The second evaluation will fail because it evaluates the whole string.
+        /// 
+        /// https://github.com/Microsoft/msbuild/issues/2259
+        /// </summary>
+        [Theory]
+        [InlineData("<Target Name=\"Build\" /><Import Project=\"$(NonExistentProperty)\" Condition=\"\'true\' == \'false\' And \'$([MSBuild]::GetDirectoryNameOfFileAbove($(NonExistentProperty), init.props))\' != \'\'\" />")]
+        [InlineData("<Target Name=\"Build\" Condition=\"\'true\' == \'false\' And \'$([MSBuild]::GetDirectoryNameOfFileAbove($(NonExistentProperty), init.props))\' != \'\'\" />")]
+        [InlineData("<Target Name=\"Build\"><Message Text=\"Build executed\" Condition=\"\'true\' == \'false\' And \'$([MSBuild]::GetDirectoryNameOfFileAbove($(NonExistentProperty), init.props))\' != \'\'\" /></Target>")]
```

Comment:
Hmm. That's way better (thanks to not crashing) but I wonder if we could do even more. Would love to see it expand the arguments to the property function even if it can't to the function itself, so something like

`[MSBuild]::GetDirectoryNameOfFileAbove(, init.props)`

Think that's doable?